### PR TITLE
feat: upgrade conference app to Ionic 7

### PR DESCRIPTION
**Pull request by sean-perkins**
This PR updates the React Conference app to the "latest", including Ionic Framework v7, React v18, Vite, Swiper 9 (using Swiper Element) and fixes bugs related to the app.

All changes:
- Added prettier to format the codebase
- Migrated from `react-scripts@3` to `vite`
  - Mapped `npm run start` to `npm run dev` to reduce churn
- Migrated from `react@17` and `react-dom@17` to `react@18` and `react-dom@18`
  - Updated type usage to use `PropsWithChildren` where appropriate
- Migrated from `swiper@7` to using CSS scroll snapping
- Updated outdated links to Github repository
- Updated outdated links to the getting started docs
- Fixed a bug where checkbox inside of filter would not toggle correctly
- Migrated form controls to simplified syntax with `label` and `label-placement`
- Migrated "error" slot content into actual error slot of form control
- Removed "dead code" (commented out code, unused imports, etc.)
- Updated to Capacitor v4

Resolves: #72, #95, #50

### How did I test this

- [x] Schedule tab loads with a list of events
  - [x] Both "All" and "Favorites" works (if you favorite an event it shows up in the segment)
- [x] Speakers tab loads with a list of speakers
  - [x] Images of speaker's avatar loads correctly
  - [x] Tapping a speaker card opens the detail view for the speaker
     - [x] Social links open to correct location
  - [x] Map tab loads (in browser requires an API key)
  - [x] About tab loads 
    - [x] Location image renders correctly in the header
    - [x] Form controls do not produce warnings
  - [x] Login page loads without warnings
  - [x] Signup page loads without warnings
  - [x] Dark mode (in the menu) updates the appearance of the application
  - [x] Tutorial renders correctly
    - [x] User can swipe to the different slides
    - [x] User can click "skip" to navigate to the main tabs view
    - [x] User can click "continue" on the last slide to navigate to the main tabs view
- [x] Tested on iPhone 14 Simulator (with Capacitor)
- [x] Tested on Pixel 5 (API 32) Simulator (with Capacitor)

### Other information

This PR includes the work done in #104.

### CODE REVIEW

**Review by liamdebeasi — COMMENTED** [review]
> Migrated from swiper@7 to swiper@9, includes moving to the Swiper Element web component

The Slides migration guide (https://ionicframework.com/docs/react/slides) recommends sticking with the Swiper React component for now. We might want to hold off on Swiper Element in React until https://github.com/nolimits4web/swiper/issues/6466 is merged.

**Comment by sean-perkins** [comment]
Updated to use CSS scrolling snapping. Following the style/implementation that is being implemented for the Angular conference app: https://github.com/ionic-team/ionic-conference-app/pull/1790

File: package.json [hunk 1]
```diff
@@ -3,31 +3,28 @@
...
+    "@capacitor/preferences": "^4.0.2",
+    "@ionic/react": "^7.0.0-rc.2",
+    "@ionic/react-router": "^7.0.0-rc.2",
     "date-fns": "^2.25.0",
     "ionicons": "^5.0.1",
-    "react": "^17.0.2",
-    "react-dom": "^17.0.2",
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
     "react-router": "^5.2.0",
```
Comment (liamdebeasi) [inline package.json:14]: Can we also update the react router deps to the latest v5? I get the following peer dependency warning:

```
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: mini-create-react-context@0.4.1
npm WARN Found: react@18.2.0
npm WARN node_modules/react
npm WARN   react@"^18.2.0" from the root project
npm WARN   6 more (@ionic/react, @ionic/react-router, ...)
npm WARN 
npm WARN Could not resolve dependency:
npm WARN peer react@"^0.14.0 || ^15.0.0 || ^16.0.0 || ^17.0.0" from mini-create-react-context@0.4.1
npm WARN node_modules/react-router/node_modules/mini-create-react-context
npm WARN   mini-create-react-context@"^0.4.0" from react-router@5.2.0
npm WARN   node_modules/react-router
npm WARN 
npm WARN Conflicting peer dependency: react@17.0.2
npm WARN node_modules/react
npm WARN   peer react@"^0.14.0 || ^15.0.0 || ^16.0.0 || ^17.0.0" from mini-create-react-context@0.4.1
npm WARN   node_modules/react-router/node_modules/mini-create-react-context
npm WARN     mini-create-react-context@"^0.4.0" from react-router@5.2.0
```

(I had to fix the same issue in the swiper migration repo)

File: src/pages/Tutorial.scss [hunk 2]
```diff
@@ -4,18 +4,37 @@
...
 
+  .slider {
+    display: grid;
+    grid-template-columns: repeat(4, 100%);
+    grid-template-rows: 1fr;
+
+    height: 100%;
+
+    overflow: scroll;
+    scroll-snap-type: x mandatory;
```
Comment (liamdebeasi) [inline src/pages/Tutorial.scss:19]: Can we add `scroll-snap-stop` you can't accidentally swipe over multiple slides: https://developer.mozilla.org/en-US/docs/Web/CSS/scroll-snap-stop

File: src/pages/SchedulePage.tsx [hunk 3]
```diff
@@ -43,61 +66,66 @@ const SchedulePage: React.FC<SchedulePageProps> = ({ favoritesSchedule, schedule
...
-          {showSearchbar &&
-            <IonSearchbar showCancelButton="always" placeholder="Search" onIonChange={(e: CustomEvent) => setSearchText(e.detail.value)} onIonCancel={() => setShowSearchbar(false)}></IonSearchbar>
-          }
+          )}
+          {!ios && !showSearchbar && <IonTitle>Schedule</IonTitle>}
+          {showSearchbar && (
+            <IonSearchbar
+              showCancelButton="always"
+              placeholder="Search"
+              onIonChange={(e: CustomEvent) => setSearchText(e.detail.value)}
```
Comment (liamdebeasi) [inline src/pages/SchedulePage.tsx:95]: Should this be `onIonInput` now? The search is only going to update on blur

File: src/pages/Login.tsx [hunk 4]
```diff
@@ -50,59 +68,74 @@ const Login: React.FC<LoginProps> = ({setIsLoggedIn, history, setUsername: setUs
...
+              <IonInput
+                label="Username"
+                labelPlacement="stacked"
+                color="primary"
+                name="username"
+                type="text"
+                value={username}
+                spellCheck={false}
+                autocapitalize="off"
+                onIonChange={(e) => setUsername(e.detail.value as string)}
```
Comment (liamdebeasi) [inline src/pages/Login.tsx:87]: Should this be onIonInput?

File: src/pages/Login.tsx [hunk 5]
```diff
@@ -50,59 +68,74 @@ const Login: React.FC<LoginProps> = ({setIsLoggedIn, history, setUsername: setUs
...
-              <IonLabel position="stacked" color="primary">Password</IonLabel>
-              <IonInput name="password" type="password" value={password} onIonChange={e => setPassword(e.detail.value!)}>
+              <IonInput
+                label="Password"
+                labelPlacement="stacked"
+                color="primary"
+                name="password"
+                type="password"
+                value={password}
+                onIonChange={(e) => setPassword(e.detail.value as string)}
```
Comment (liamdebeasi) [inline src/pages/Login.tsx:106]: Should this be onIonInput?

File: src/pages/Support.tsx [hunk 6]
```diff
@@ -39,45 +54,54 @@ const Support: React.FC<SupportProps> = () => {
...
+              <IonTextarea
+                label="Enter your support message below"
+                labelPlacement="stacked"
+                color="primary"
+                name="message"
+                value={message}
+                spellCheck={false}
+                autocapitalize="off"
+                rows={6}
+                onIonChange={(e) => setMessage(e.detail.value!)}
```
Comment (liamdebeasi) [inline src/pages/Support.tsx:73]: Should this be onIonInput?

**Review by liamdebeasi — CHANGES_REQUESTED** [review]

**Review by brandyscarney — CHANGES_REQUESTED** [review]
### Differences from Angular

I noticed that the react conference app behaves differently than the angular version in certain cases. Maybe we can make a tech debt ticket for these if we don't want to update them now? 

`ionic-conference-app` -> shows toast when favoriting
`ionic-react-conference-app` -> shows alert when favoriting

The alert text when you attempt to remove a favorite or add a favorite that already exists is also different:

Frame 1

It also does not [close the open sliding item](https://github.com/ionic-team/ionic-conference-app/blob/5ad793c582221fc8f4f2ea7a7dbf4788710798a3/src/app/pages/schedule/schedule.ts#L49) when switching between all & favorites.

### Login Bugs

When logging in (Login page) after I enter name & password and click "Login" it tells me "Password is required" until I press login again. I think this is related to the `ionChange` Liam mentioned. Support & Signup do the same thing. 

The app is continuously freezing when I click on Logout:

Screenshot 2023-03-29 at 3 47 35 PM

I have to kill the tab and open a new one, refresh doesn't work.

**Comment by sean-perkins** [comment]
Ok - I addressed all of the requested changes and differences in behavior we've acknowledge so far with the React and Angular conference apps (there is likely more).

File: src/pages/Signup.tsx [hunk 7]
```diff
@@ -50,62 +68,75 @@ const Login: React.FC<LoginProps> = ({setIsLoggedIn, history, setUsername: setUs
...
+              <IonInput
+                label="Username"
+                labelPlacement="stacked"
+                color="primary"
+                name="username"
+                type="text"
+                value={username}
+                spellCheck={false}
+                autocapitalize="off"
+                onIonChange={(e) => {
```
Comment (brandyscarney) [inline src/pages/Signup.tsx:87]:
```suggestion
                onIonInput={(e) => {
```

File: src/pages/Signup.tsx [hunk 8]
```diff
@@ -50,62 +68,75 @@ const Login: React.FC<LoginProps> = ({setIsLoggedIn, history, setUsername: setUs
...
-                setPasswordError(false);
-              }}>
+              <IonInput
+                label="Password"
+                labelPlacement="stacked"
+                color="primary"
+                name="password"
+                type="password"
+                value={password}
+                onIonChange={(e) => {
```
Comment (brandyscarney) [inline src/pages/Signup.tsx:109]:
```suggestion
                onIonInput={(e) => {
```

**Review by brandyscarney — CHANGES_REQUESTED** [review]

**Comment by brandyscarney** [comment]
All the fixes look good! Just found two missed `ionChange` events on the signup page.

**Comment by sean-perkins** [comment]
Thank you for catching that!

**Review by liamdebeasi — APPROVED** [review]

**Pull request merged**